test/e2e/common/expansion.go

```diff
@@ -306,6 +308,317 @@ var _ = framework.KubeDescribe("Variable Expansion", func() {
+		    Description: Verify that a subpath expansion can be modified during the lifecycle of a container.
+			1.	valid subpathexpr starts a container running
+			2.	test for valid subpath writes
+			3.	container restarts
+			4.  delete cleanly
+
+	*/
+	It("should succeed in altering lifecycle of container [Feature:VolumeSubpathEnvExpansion][NodeAlphaFeature:VolumeSubpathEnvExpansion][Slow]", func() {
+
+		suffix := string(uuid.NewUUID())
+		podName := fmt.Sprintf("var-expansion-%s", suffix)
+		containerName := "dapi-container"
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   podName,
+				Labels: map[string]string{"name": podName},
+			},
+			Spec: v1.PodSpec{
+				InitContainers: []v1.Container{
+					{
+						Name:    fmt.Sprintf("init-volume-%s", suffix),
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "mkdir -p /volume_mount/foo; touch /volume_mount/foo/test.log"},
+						Env: []v1.EnvVar{
+							{
+								Name:  "POD_NAME",
+								Value: "foo",
+							},
+						},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:        "workdir1",
+								MountPath:   "/subpath_mount",
+								SubPathExpr: "$(POD_NAME)",
+							},
+							{
+								Name:      "workdir2",
+								MountPath: "/volume_mount",
+							},
+						},
+					},
+				},
+
+				Containers: []v1.Container{
+					{
+						Name:    containerName,
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "tail -f /dev/null"},
+						Env: []v1.EnvVar{
+							{
+								Name:  "POD_NAME",
+								Value: "foo",
+							},
+						},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:        "workdir1",
+								MountPath:   "/subpath_mount",
+								SubPathExpr: "$(POD_NAME)",
+							},
+							{
+								Name:      "workdir2",
+								MountPath: "/volume_mount",
+							},
+						},
+					},
+					{
+						Name:    containerName + "2",
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "tail -f /dev/null"},
+						Env: []v1.EnvVar{
+							{
+								Name:  "POD_NAME",
+								Value: "foo",
+							},
+						},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:        "workdir1",
+								MountPath:   "/subpath_mount",
+								SubPathExpr: "$(POD_NAME)",
+							},
+							{
+								Name:      "workdir2",
+								MountPath: "/volume_mount",
+							},
+						},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+				Volumes: []v1.Volume{
+					{
+						Name: "workdir1",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: "/tmp"},
+						},
+					},
+					{
+						Name: "workdir2",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: "/tmp"},
+						},
+					},
+				},
+			},
+		}
+
+		By("waiting for pod and container restart")
+		subpath.WaitForPodContainerRestart(f, pod, "/volume_mount/foo/test.log")
```

Comment:
I think that whole test is self contained. It creates the pod, checks the file rewrites and deletes the pod. So I dont believe there is a way of providing additional checks to it without duplicating it

Reply:
We may have refactor that method into smaller parts so that we can inject our test case into it.  The differences between the env var tests and the normal subpath tests is that the env var tests also need to:

* Change the env var after the pod is started, before triggering a container restart
* Validate that the subpath directory did not change after container restart

Reply:
I reverted the subpath inclusion suite and added a container restart section. There's an element if duplication but allows freedom to inject test cases rather than a potential break to subpath suites